sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,275 @@
+# -*- coding: utf-8 -*-
+"""Recommendation system generator.
+
+Generate recommendations based on TF-IDF representation and a KNN model.
+"""
+# Author: Arturo Amor
+# License: 3-clause BSD
+
+
+import numbers
+import re
+from collections import defaultdict
+from pathlib import Path
+
+from sphinx.errors import ConfigError
+from .backreferences import (
+    _thumbnail_div,
+    THUMBNAIL_PARENT_DIV,
+    THUMBNAIL_PARENT_DIV_CLOSE,
+)
+from .py_source_parser import split_code_and_text_blocks
+from .gen_rst import extract_intro_and_title
+from .utils import _replace_md5
+
+
+class ExampleRecommender:
+    """Compute content-based KNN-TF-IFD recommendation system.
+
+    Parameters
+    ----------
+    n_examples : int, default=5
+        Number of most relevant examples to display.
+
+    tokenizer : {"raw", "backrefs"}, default="raw"
+        The type of tokenizer to use. If "raw", the raw text is used as
+        tokens. If "backrefs", the list of sphinx-gallery backreferences is
+        used as tokens.
+
+    Attributes
+    ----------
+    file_names_ : list of str
+        The list of file names used for computing the similarity matrix.
+        The recommended examples are chosen among this list.
+
+    similarity_matrix_ : dense matrix
+        Fitted matrix of pairwise cosine similarities.
+    """
+
+    def __init__(self, *, n_examples=5, tokenizer="raw"):
+        try:
+            import numpy as np
+
+            self.np = np
+        except ImportError:
+            raise ConfigError("gallery_conf['recommender'] requires numpy")
+        self.n_examples = n_examples
+        self.tokenizer = tokenizer
+
+    def token_freqs(self, doc):
+        """Extract a dict mapping raw tokens from doc to their occurrences."""
+        token_generator = (tok.lower() for tok in re.findall(r"\w+", doc))
+        return self.dict_freqs(token_generator)
+
+    @staticmethod
+    def dict_freqs(doc):
+        """Extract a dict mapping list of tokens to their occurrences."""
+        freq = defaultdict(int)
+        for tok in doc:
+            freq[tok] += 1
+        return freq
+
+    def dict_vectorizer(self, data):
+        """Convert a dictionary of feature arrays into a matrix.
+
+        Parameters
+        ----------
+        data : list of dict
+            An iterable of dictionaries of feature arrays, where each key
+            corresponds to a feature name, and each value is an array of
+            feature values.
+
+        Returns
+        -------
+        X : ndarray of shape (n_samples, n_features)
+            A matrix of occurrences where n_samples is the number of samples in
+            the dataset and n_features is the total number of features across
+            all samples.
+        """
+        np = self.np
```

Comment:
Someone just jumping into this bit of code actually has to do some thinking. We can reduce the burden just by doing an explicit nested import each time
```
```suggestion
        import numpy as np
```